type4py/gen_cluster.py

```diff
@@ -0,0 +1,317 @@
+        idx, dist = indexed_knn.get_nns_by_vector(embed_vec, k, include_distances=True)
+        pred_idx_scores = compute_types_score(dist, idx, type_space_labels)
+
+        pred_types.append({'original_type': types_embed_labels[i], 'predictions': pred_idx_scores,
+                           'task': find_pred_task(i),
+                           'is_parametric': bool(re.match(r'(.+)\[(.+)\]', types_embed_labels[i]))})
+
+        # pred_types_embed.append([i for i, s in pred_idx_scores])
+        # pred_types_score.append(pred_idx_scores)
+
+    return pred_types
+
+
+def build_type_clusters(model, output_path, train_data_loader: DataLoader, valid_data_loader: DataLoader, type_vocab: set,
+                        exist_index: str, exist_emd: str):
+    logger.info("Type Cluster building begin...")
+    computed_embed_labels = []
+    annoy_idx = AnnoyIndex(model.output_size, 'euclidean')
+    loaded_idx = AnnoyIndex(model.output_size, 'euclidean')
+    curr_idx = 0
+
+    if exist_index is not None:
+        loaded_idx.load(join(output_path, exist_index))
+        curr_idx = loaded_idx.get_n_items()
+        for i in range(loaded_idx.get_n_items()):
+            item_vector = loaded_idx.get_item_vector(i)
+            annoy_idx.add_item(i, item_vector)
+
+    if exist_emd is not None:
+        embedd_labels = np.load(join(output_path, exist_emd)).tolist()
+        computed_embed_labels.extend(embedd_labels)
+
+    for _, (a, _, _) in enumerate(
+            tqdm(train_data_loader, total=len(train_data_loader), desc="Computing Type Clusters - Train set")):
+        model.eval()
+        with torch.no_grad():
+            output_a = model(*(s.to(DEVICE) for s in a[0]))
+            lables = a[1].data.cpu().numpy()
+            # computed_embed_labels.append(lables)
+            for i, v in enumerate(output_a.data.cpu().numpy()):
+                if lables[i] in type_vocab:
+                    annoy_idx.add_item(curr_idx, v)
+                    computed_embed_labels.append(lables[i])
+                    curr_idx += 1
+
+    for _, (a, _, _) in enumerate(
+            tqdm(valid_data_loader, total=len(valid_data_loader), desc="Computing Type Clusters - Valid set")):
+        model.eval()
+        with torch.no_grad():
+            output_a = model(*(s.to(DEVICE) for s in a[0]))
+            lables = a[1].data.cpu().numpy()
+            # computed_embed_labels.append(a[1].data.cpu().numpy())
+            for i, v in enumerate(output_a.data.cpu().numpy()):
+                if lables[i] in type_vocab:
+                    annoy_idx.add_item(curr_idx, v)
+                    computed_embed_labels.append(lables[i])
+                    curr_idx += 1
+
+    annoy_idx.build(KNN_TREE_SIZE)
+    # annoy_idx.
+    return annoy_idx, np.array(computed_embed_labels)  # np.hstack(computed_embed_labels)
+
+
+def compute_type_embed_batch(model, data_loader: DataLoader, pca: PCA = None) -> Tuple[np.array, np.array]:
```

Comment:
If the logic of the method is the same as the one in the `predict` module. Reuse the one from `predict`.